File: benchmark/mandelbrot/mandelbrot.c
#include <stdio.h>

static int mandelbrot(int width, int height, int max_iter) {
    int inside = 0;

    for (int y = 0; y < height; y++) {
        for (int x = 0; x < width; x++) {
            double cr = ((double)x / (double)width) * 3.5 - 2.5;
            double ci = ((double)y / (double)height) * 2.0 - 1.0;
            double zr = 0.0;
            double zi = 0.0;
            int iter = 0;

            while (((zr * zr) + (zi * zi)) <= 4.0 && iter < max_iter) {
                double next_zr = ((zr * zr) - (zi * zi)) + cr;
                zi = ((2.0 * zr) * zi) + ci;
                zr = next_zr;
                iter++;
            }

            if (iter == max_iter) {
                inside++;
            }
        }
    }

    return inside;
}

int main() {
    int points = mandelbrot(800, 600, 80);
    printf("Mandelbrot points: %d\n", points);
    return 0;
}

File: benchmark/mandelbrot/mandelbrot.js
function mandelbrot(width, height, maxIter) {
  let inside = 0;

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const cr = (x / width) * 3.5 - 2.5;
      const ci = (y / height) * 2.0 - 1.0;
      let zr = 0.0;
      let zi = 0.0;
      let iter = 0;

      while ((zr * zr) + (zi * zi) <= 4.0 && iter < maxIter) {
        const nextZr = ((zr * zr) - (zi * zi)) + cr;
        zi = ((2.0 * zr) * zi) + ci;
        zr = nextZr;
        iter++;
      }

      if (iter === maxIter) {
        inside++;
      }
    }
  }

  return inside;
}

console.log(`Mandelbrot points: ${mandelbrot(800, 600, 80)}`);
